components/style/values/computed/background.rs

```diff
@@ -29,59 +27,6 @@ impl BackgroundSize {
 
 impl RepeatableListAnimatable for BackgroundSize {}
 
-impl ToAnimatedZero for BackgroundSize {
-    #[inline]
-    fn to_animated_zero(&self) -> Result<Self, ()> { Err(()) }
-}
-
-impl ToAnimatedValue for BackgroundSize {
-    type AnimatedValue = Self;
-
-    #[inline]
-    fn to_animated_value(self) -> Self {
-        self
-    }
-
-    #[inline]
-    fn from_animated_value(animated: Self::AnimatedValue) -> Self {
-        use values::computed::{Length, Percentage};
-        let clamp_animated_value = |value: LengthOrPercentageOrAuto| -> LengthOrPercentageOrAuto {
-            match value {
```

Comment:
So, as we discussed on IRC, this works, but this removes the clamping behavior on BackgroundSize... I just pushed this to try (https://treeherder.mozilla.org/#/jobs?repo=try&revision=d7479f22e46b38c6103a46045e1bf83186927b18&selectedJob=149390573) and it looks green.

@BorisChiou, you added this code in https://github.com/servo/servo/commit/ebedea58608b8326129b987caad5c4674fbabad2 (#17783).

Can we remove this? If not, could you write guidance on how to write a test for this?

Reply:
Unfortunately, Gecko doesn't test this [1]. I added this because we may use a negative cubic timing function for animations/transitions. If you want to verify your patch, maybe try this https://jsfiddle.net/borischiou/w5yqs437/. If the console output is "0px 0px", it works. However, if you see something like "calc(-68.3px) calc(-68.3px)", please fix it. :)

[1] https://searchfox.org/mozilla-central/rev/477ac066b565ae0eb3519875581a62dfb1430e98/layout/style/test/test_transitions_per_property.html#109-113

Reply:
And I guess you may get the negative value (i.e. unclamped value) if animation_value_type == "ComputedValue" for ```background-size```. ```mask-size``` has the similar problem, I think. Therefore, It's better not to remove this.

Reply:
@BorisChiou After trying to access the [jsfiddle link](https://jsfiddle.net/borischiou/w5yqs437/), I found it logged with `0px 0px` in my local. However, it got an error of `ERROR:script::dom::bindings::error: Error at https://jsfiddle.net/js/_dist-editor.js?update_24_11_2017_2:1:228763 Ao(...).getBoundingClientRect is not a function` so I cannot see the jsfiddle UI.
Not sure if it means updated in this PR works?

Reply:
Really? Oh, I used Firefox to verify this jsfiddle test. Sorry, I will try Servo browser again.

Reply:
@BorisChiou Yeah! I built with this branch and try to run `./mach run https://jsfiddle.net/borischiou/w5yqs437/` locally. Then, I saw `0px 0px` in my terminal console. Please help me verify it, thanks!

Reply:
Looks like there are some bugs so Servo cannot run jsfiddle perfectly. I have a new test file for you. Please run it locally. [link](https://pastebin.mozilla.org/9074070). Servo doesn't flush the computed style correctly immediately, so we have to wait for a while (so I use 500ms). Please check this result after waiting for 500ms.

Another way to debug this is [1]. We call ```update()``` during the animation, and the ```value``` here shouldn't be negative because it is a size.

[1] https://searchfox.org/mozilla-central/rev/ba2b0cf4d16711d37d4bf4d267b187c9a27f6638/servo/components/style/properties/helpers/animated_properties.mako.rs#278

Reply:
@BorisChiou Ok, thanks! I'll try it!

Reply:
@BorisChiou After running the html locally, I got negative size with `after 500ms: -68.7292px -68.7292px` in console. I'm going to see what I'll log with `println!`.

BTW, I cannot see the reference link 😂

Reply:
@BorisChiou cc @emilio 
After running the html Boris provided locally, here's part of the result.

```
T([Explicit { width: 0px, height: 0px }])
T([Explicit { width: -64.5544px, height: -64.5544px }])
T([Explicit { width: -64.903px, height: -64.903px }])
T([Explicit { width: -65.3792px, height: -65.3792px }])
T([Explicit { width: -65.8209px, height: -65.8209px }])
T([Explicit { width: -66.2986px, height: -66.2986px }])
T([Explicit { width: -66.6553px, height: -66.6553px }])
T([Explicit { width: -67.0509px, height: -67.0509px }])
T([Explicit { width: -67.3707px, height: -67.3707px }])
T([Explicit { width: -67.673px, height: -67.673px }])
T([Explicit { width: -67.8893px, height: -67.8893px }])
T([Explicit { width: -68.1136px, height: -68.1136px }])
T([Explicit { width: -68.2917px, height: -68.2917px }])
T([Explicit { width: -68.4515px, height: -68.4515px }])
T([Explicit { width: -68.5687px, height: -68.5687px }])
T([Explicit { width: -68.6605px, height: -68.6605px }])
T([Explicit { width: -68.7209px, height: -68.7209px }])
T([Explicit { width: -68.747px, height: -68.747px }])
T([Explicit { width: -68.75px, height: -68.75px }])
T([Explicit { width: -68.75px, height: -68.75px }])
T([Explicit { width: -58.5046px, height: -58.5046px }])
T([Explicit { width: -58.3998px, height: -58.3998px }])
T([Explicit { width: -58.353px, height: -58.353px }])
T([Explicit { width: -58.2164px, height: -58.2164px }])
after 500ms: -68.75px -68.75px
T([Explicit { width: -57.9915px, height: -57.9915px }])
T([Explicit { width: -57.9915px, height: -57.9915px }])
T([Explicit { width: -57.5878px, height: -57.5878px }])
T([Explicit { width: -56.9537px, height: -56.9537px }])
T([Explicit { width: -56.2757px, height: -56.2757px }])
......
T([Explicit { width: 99.8221px, height: 99.8221px }])
T([Explicit { width: 99.9465px, height: 99.9465px }])
T([Explicit { width: 99.9945px, height: 99.9945px }])
T([Explicit { width: 100px, height: 100px }])
T([Explicit { width: 100px, height: 100px }])
```

It starts from `0px 0px`; then get negative numbers immediately.
Then, it will add the number from `-64px` until `100px`.

Reply:
Yes. I guess it starts from ```0px 0px``` because its style is ```0px 0px``` in the beginning. After starting to run the animation, we replace the style value with the interpolated animation value according to the timing function, i.e. ```width: -64.5544px, height: -64.5544px```, and then keep running. Therefore, here, we have to make sure both ```width``` and ```height``` are always positive during the animation [1].

> I cannot see the reference link 

How about dxr? :)
[1] https://dxr.mozilla.org/servo/rev/b56d7b8070e3280a6dc43463b110e1a1cb4ea305/components/style/properties/helpers/animated_properties.mako.rs#274-278

Reply:
@BorisChiou @emilio Thanks! Now, I'm going to fix the `ToComputedValue` trait for `BackgroundSize`.

> How about dxr? :)

Yeah! I can see the code in `dxr`!

Reply:
@BorisChiou @emilio I just updated it in the second commit of this PR. After `impl ToComputedValue for BackgroundSize`, I got the result of running servo with the provided html as below.

```
T([Explicit { width: 0px, height: 0px }])
T([Explicit { width: -62.8527px, height: -62.8527px }])
T([Explicit { width: -63.4729px, height: -63.4729px }])
.........
T([Explicit { width: -68.4964px, height: -68.4964px }])
T([Explicit { width: -68.3639px, height: -68.3639px }])
T([Explicit { width: -68.1362px, height: -68.1362px }])
after 500ms: 0px 0px
T([Explicit { width: -68.0298px, height: -68.0298px }])
T([Explicit { width: -67.962px, height: -67.962px }])
T([Explicit { width: -67.962px, height: -67.962px }])
.........
T([Explicit { width: -5.40848px, height: -5.40848px }])
T([Explicit { width: -4.25273px, height: -4.25273px }])
T([Explicit { width: -2.9012px, height: -2.9012px }])
T([Explicit { width: -1.80122px, height: -1.80122px }])
T([Explicit { width: -0.4725px, height: -0.4725px }])
T([Explicit { width: 0.66849px, height: 0.66849px }])
.........
T([Explicit { width: 100px, height: 100px }])
```

Could you help me to verify these piece of code? Thanks!

Reply:
Hmm... I got `after 500ms: -68.7001px -68.7001px` when I ran it locally second time.
I'll check it more detail tomorrow.

Reply:
@BorisChiou `mask-size` was already bogus here, wasn't it? In particular, it was using `animation_value_size="ComputedValue"`.

Could we get that stuff tested pretty please?

Reply:
Yes..we have to do that. We have to update the test cases in test_transitions_per_property.html for background-size and mask-size (so at least we test it on Firefox), and fix the problem of mask-size. I have to file a bug for it.

Reply:
@BorisChiou Thanks for the update! I think we can merge this PR and let's fix the issue in another PR?

cc @emilio

Reply:
But the problem is: this PR makes background-size be buggy for animation. At least now we have to make sure background-size is correct for animation.